### PR TITLE
Add ServerlessScope.SERVERLESS_RESTRICTION constant

### PR DESCRIPTION
**Pull request by pgomulka**
This commit adds a ServerlessScope.SERVERLESS_RESTRICTION constant that will be used as a marker on requests from the non-operator users in order to restrict the response

- Have you signed the [contributor license agreement](https://www.elastic.co/contributor-agreement)?
- Have you followed the [contributor guidelines](https://github.com/elastic/elasticsearch/blob/main/CONTRIBUTING.md)?
- If submitting code, have you built your formula locally prior to submission with `gradle check`?
- If submitting code, is your pull request against main? Unless there is a good reason otherwise, we prefer pull requests against main and will backport as needed.
- If submitting code, have you checked that your submission is for an [OS and architecture that we support](https://www.elastic.co/support/matrix#show_os)?
- If you are submitting this code for a class then read our [policy](https://github.com/elastic/elasticsearch/blob/main/CONTRIBUTING.md#contributing-as-part-of-a-class) for that.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/rest/ServerlessScope.java [hunk 1]
```diff
@@ -22,4 +22,9 @@
 @Target(ElementType.TYPE)
 public @interface ServerlessScope {
     Scope value();
+
+    /**
+     * A value used when restricting a response of a serverless endpoints.
+     */
+    String SERVERLESS_RESTRICTION = "serverless";
```
Comment (rjernst) [inline server/src/main/java/org/elasticsearch/rest/ServerlessScope.java:29]: This seems to be unrelated to the PR title?

File: modules/rest-root/src/main/java/org/elasticsearch/rest/root/MainRequest.java [hunk 2]
```diff
@@ -18,7 +18,7 @@ public class MainRequest extends ActionRequest {
 
     public MainRequest() {}
 
-    MainRequest(StreamInput in) throws IOException {
+    public MainRequest(StreamInput in) throws IOException {
```
Comment (rjernst) [inline modules/rest-root/src/main/java/org/elasticsearch/rest/root/MainRequest.java:21]: This also seems to be unrelated to the PR title?

File: server/src/main/java/org/elasticsearch/rest/RestRequest.java [hunk 3]
```diff
@@ -48,6 +48,7 @@
 public class RestRequest implements ToXContent.Params {
 
     public static final String RESPONSE_RESTRICTED = "responseRestricted";
+
```
Comment (rjernst) [inline server/src/main/java/org/elasticsearch/rest/RestRequest.java:51]: whitespace change?

File: build-tools-internal/src/main/java/org/elasticsearch/gradle/internal/precommit/JavaModulePrecommitTask.java [hunk 4]
```diff
@@ -116,8 +116,11 @@ private void checkModuleVersion(ModuleReference mref) {
 
     private void checkModuleNamePrefix(ModuleReference mref) {
         getLogger().info("{} checking module name prefix for {}", this, mref.descriptor().name());
-        if (mref.descriptor().name().startsWith("org.elasticsearch.") == false) {
-            throw new GradleException("Expected name starting with \"org.elasticsearch.\", in " + mref.descriptor());
+        if (mref.descriptor().name().startsWith("org.elasticsearch.") == false
```
Comment (rjernst) [inline build-tools-internal/src/main/java/org/elasticsearch/gradle/internal/precommit/JavaModulePrecommitTask.java:119]: Rather than hardcoding this, I wonder if it is something serverless should pass in (say via a flag or some other Gradle mechanism for composite builds), so that we don't start allowing co.elasticsearch to leak into the main repo. But I leave this up to @mark-vieira.

Comment (mark-vieira) [reply build-tools-internal/src/main/java/org/elasticsearch/gradle/internal/precommit/JavaModulePrecommitTask.java:119]: Yeah, I think we should make this task more extensible. This same task also enforces module version, which might not always be applicable. If this is blocking other work we can probably disable this in serverless for now. The only check I think would be worthwhile keeping there is the services bit.

Comment (pgomulka) [reply build-tools-internal/src/main/java/org/elasticsearch/gradle/internal/precommit/JavaModulePrecommitTask.java:119]: I am creating a new module in serverless where I wanted to use a name of a module with `co.elasticsearch.etc.etc..`
if we can disable it, it would work for me

File: server/src/main/java/org/elasticsearch/rest/ServerlessScope.java (hunk 1 above)
Comment (pgomulka) [reply server/src/main/java/org/elasticsearch/rest/ServerlessScope.java:29]: updated the PR

**Review by jakelandis — APPROVED** [review]
LGTM

**Comment by pgomulka** [comment]
@elasticmachine update branch
